Test: Add Custom Junit reporter

eloycoto opened the pull request:
Hi, 

This PR introduces a new way of report some checks on a failed test. The idea is to provide more context to users when a test failed, so this introduce a Cilium status summary. how many policies are loaded, endpoints and their policy enforcement, cilium status, controllers and if any deadlock,panic or how many times the error/warning/context deadline are present in the logs. 

The summary of changes: 
- New Junit reporter that support system-err and system-out
- Added the Checks buffer per test and dump to stdout using <checks> labels. 
- Extend ValidateNoErrorsOnLogs to check logs and make warnings. 
- Added CiliumCheckReport

The output on Jenkins will be similar to this: 

```

Number of "context deadline exceeded" in logs: 0
Number of "level=error" in logs: 0
Number of "level=warning" in logs: 0
Cilium pods: [cilium-8nrxq cilium-czqxs]
Netpols loaded: default::second
CiliumNetworkPolicies loaded: default::rule1
Cilium CEP status:
        cilium-health-k8s2=> none
        kube-dns-f4d788bb7-htj44=> none
        microscope=> none
        app1-78b4c7b56c-6jfcd=> ingress
        app1-78b4c7b56c-gdg4v=> ingress
        app2-6db56c7d8f-2hbbv=> none
        app3-6d745c45f6-8dldc=> none
Cilium agent "cilium-8nrxq": Status: Ok  Health: Ok Nodes "k8s2 k8s1" ContinerRuntime: Ok Kubernetes: Ok KVstore: Ok Controllers: Total 39 Failed 0
Cilium agent "cilium-czqxs": Status: Ok  Health: Ok Nodes "k8s2 k8s1" ContinerRuntime: Ok Kubernetes: Ok KVstore: Ok Controllers: Total 27 Failed 0

```
<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/4823)
<!-- Reviewable:end -->

eloycoto committed:
CI: Custom junit reporter

Dump Ginkgo Junit reporter and add custom filters on stdout/stderr in
failing test.

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

eloycoto committed:
Test: Fix Deadlock check

Use previous and current logs in the deadlock detection.
Deleted unused fuction helper.

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

eloycoto committed:
Test: Add Checks output on Jenkins CI XML.

Add Checks output to be rendered on Junit standar output.

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

eloycoto committed:
Test: Add Check results on Deadlock helper

Added the new check validation on Deadlock helper in case that was found
it.

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

eloycoto committed:
Test: Add CiliumCheckReport on failed test.

Added a new function that will print check reports on failing test to
provide more infomation to developers about what failed.

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

eloycoto committed:
Test: Added the number of issues on report.

Add the number of context-deadline, error messages and warning messages
in the test results to provide more context to users.

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

eloycoto committed:
Test: Fix lint issues

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

raybejjani commented on test/helpers/kubectl.go:
```diff
@@ -912,37 +911,87 @@ func (kub *Kubectl) CiliumReport(namespace string, commands ...string) {
+		"%s get cep -o jsonpath='%s' --all-namespaces",
+		KubectlCmd, cepFilter))
+
+	fmt.Fprintf(CheckLogs, "Cilium CEP status:\n")
```
Since it's the policy enforcement value, it's more clear to say "Endpoint Policy Enforcement" instead of CEP status. We can also say (from CEP) at the end if you think that part is important to keep.

raybejjani commented on test/helpers/kubectl.go:
```diff
@@ -912,37 +911,87 @@ func (kub *Kubectl) CiliumReport(namespace string, commands ...string) {
+
+	fmt.Fprintf(CheckLogs, "Cilium CEP status:\n")
+	for pod, policy := range cepStatus.KVOutput() {
+		fmt.Fprintf(CheckLogs, "\t%s=> %s\n", pod, policy)
```
It might also be better to print these with a tabwriter(https://golang.org/pkg/text/tabwriter/)  so they align better.

raybejjani commented on test/ginkgo-ext/junit_reporter.go:
```diff
@@ -0,0 +1,197 @@
+	case types.SpecStatePanicked:
+		return "Panic"
+	default:
+		return ""
```
Should the default say "unknown" so it's clear that it wasn't handled here? I'm not sure if blank means something specfic, however.

raybejjani commented on test/ginkgo-ext/junit_reporter.go:
```diff
@@ -0,0 +1,197 @@
+	}
+}
+
+func reportChecks(output string) (string, string) {
```
Could you add a comment explaining all the little things this is doing (and maybe why? how it fits into the larger processing)? It will probably not be clear in the future, especially if the XML format changes and we need to change this.

raybejjani commented on test/ginkgo-ext/junit_reporter.go:
```diff
@@ -0,0 +1,197 @@
+
+/*
+
+JUnit XML Reporter for Ginkgo
```
Maybe explicitly say here that it's a fork, so no one thinks we just vendored it in a funny way.

eloycoto replied:
Ir was already in there, see a bit below ;-)

ianvernon replied:
Given that this is a fork, it would make it a lot easier to review if the initial code was copied into `ginkgo-ext` (i.e., no functional change at all in the code, just the import dependency  in `test_suite_test.go` , and then you patched the changes on top of it). 

Also, are there any copyright concerns / copyright that need to be appended to the top of the file?

raybejjani commented on test/helpers/cons.go:
```diff
@@ -29,14 +31,18 @@ var (
+	// CheckLogs new buffer where all the warnings and checks that happens
+	// during the test are saved. This buffer will be printed in the test
+	// output inside <checks> labels.
+	CheckLogs = ginkgoext.NewWriter(new(bytes.Buffer))
```
"CheckLogs _creates a_ new buffer" is more clear, maybe?

Maybe also note that it isn't thread safe and so parallel tests aren't possible.

raybejjani commented on test/helpers/kubectl.go:
```diff
@@ -912,37 +912,91 @@ func (kub *Kubectl) CiliumReport(namespace string, commands ...string) {
 
-// CheckLogsForDeadlock checks if the logs for Cilium log messages that signify
-// that a deadlock has occurred.
-func (kub *Kubectl) CheckLogsForDeadlock() {
```
Was this function never called? grep shows me some places it's used but I guess it's another version of it that isn't on the `Kubectl` type?

eloycoto replied:
No for Kubectl ;-)

raybejjani commented on test/helpers/kubectl.go:
```diff
@@ -912,37 +912,91 @@ func (kub *Kubectl) CiliumReport(namespace string, commands ...string) {
 }
 
+// CiliumCheckReport prints a few checks on the Junit output to provide more context to users.
+func (kub *Kubectl) CiliumCheckReport() {
```
It might be nice to list the things we're printing here, since the code makes it hard to see because most of it is collecting and formatting.

raybejjani reviewed (approved):
LGTM

ianvernon commented on test/helpers/kubectl.go:
```diff
@@ -917,32 +916,22 @@ func (kub *Kubectl) CiliumReport(namespace string, commands ...string) {
+	var logs string
 	cmd := fmt.Sprintf("%s -n %s logs --timestamps=true -l k8s-app=cilium --since=%vs",
 		KubectlCmd, KubeSystemNamespace, duration.Seconds())
 	res := kub.Exec(fmt.Sprintf("%s --previous", cmd), ExecOptions{SkipLog: true})
```
Only check the previous logs for each individual pod if the restart count is nonzero for the given pod.

eloycoto replied:
The issue here is that we use `-l k8s-app=cilium` so we don't know the pod at all, if we want to add the pod it'll be more calls and I don't really want to do that, each call will be executed after each test, and that will take time and the time of the test will be increased.

ianvernon replied:
OK

ianvernon replied:
OK

ianvernon replied:
OK

ianvernon replied:
Sorry, GitHub wasn't showing the comment as being added, I am 3x OK with this ;)

ianvernon reviewed (approved):
Nothing blocking

eloycoto committed:
CI: Import Ginkgo Junit reporter to the project

Import the junit reporter from Ginkgo to be able to update with custom
methods in the future.

Project: https://github.com/onsi/ginkgo
File: reporters/junit_reporter.go
Commit: 3774a09d95489ccaa16032e0770d08ea77ba6184

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>
